Implement front and back camera switching

cheukchuen opened the pull request:
## Purpose
<!-- Describe the intention of the changes being proposed. What problem does it solve or functionality does it add? -->
* Enable front and back camera switching for local participant

## Does this introduce a breaking change?
<!-- Mark one with an "x". -->
```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What to Check
Verify that the following are valid
* Front and back camera switching upon tapping the camera switch button
* The camera switch button should be slightly smaller and in the top right for the detached (PIP) local participant view during 1:1
* The camera switch button should be slightly larger and in the middle right for the standard local participant view for all other calls

cheukchuen committed:
Implement front and back camera switching

jimchou-dev commented on AzureCalling/Views/ParticipantView.swift:
```diff
@@ -105,9 +108,44 @@ class ParticipantView: UIView {
+        let buttonLength: CGFloat
+        let verticalConstraint: NSLayoutConstraint
+
+        if isOneOnOne {
+            // Top right (smaller) camera switch for detached local participant view
+            buttonLength = 24
+            verticalConstraint = switchCameraButton.topAnchor.constraint(equalTo: containerView.topAnchor, constant: 4)
+        } else {
+            // Middle right camera switch for standard local participant view
+            buttonLength = 30
+            verticalConstraint = switchCameraButton.centerYAnchor.constraint(equalTo: containerView.centerYAnchor)
+        }
```
The middle right camera is to prevent the camera switch button to be covered by the InfoHeader when the caller is the only user in the call and have camera turned on.

When the call grid is is in PIP the switchCameraButton is constraint to the top-right corner. When call view grid is 2x2 or 3x2, the switchCamera Button should also be on the top-right corner. Probably can do a check if callingContext.displayedRemoteParticipants is not empty.

cheukchuen replied:
I've thought about this as well, but decided to keep the button on middle right to be consistent on all views except for PIP. It depends on how we want the design to be.

@Gemakk how did you implement this in Android? Do you think we should keep it top-right corner for 2x2 and 3x2 as well?

Gemakk replied:
Default is right.  Top right only during 1:1

Gemakk reviewed (dismissed):
Functionality fully works as tested.  I noticed specific emulator bugs, but I also verified that they're unrelated to this specific PR.

AmyL219 commented on AzureCalling/External/Calling/CallingContext.swift:
```diff
@@ -226,6 +226,27 @@ class CallingContext: NSObject {
+                    return
+                }
+                print("Camera switch successful")
+                completionHandler(.success(()))
```
if there is no flipCamera available, we still need to send out a completionHandler(.failure)

cheukchuen replied:
Good point, we still need to handle all cases even though it's unlikely to happen. `.failure` does not accept optional Error value though.

AmyL219 commented on AzureCalling/Views/ParticipantView.swift:
```diff
@@ -105,9 +108,44 @@ class ParticipantView: UIView {
 
+    func updateCameraSwitch(isOneOnOne: Bool) {
+        // Skip update if display mode is the same
+        guard detachMode != isOneOnOne else {
```
Could `detachMode` and `isOneOnOne` be the same name?

cheukchuen replied:
Sure. I could also use an alias for the parameter: `updateCameraSwitch(isOneOnOne detachMode: Bool)`

AmyL219 commented on AzureCalling/External/Calling/CallingContext.swift:
```diff
@@ -334,9 +355,15 @@ class CallingContext: NSObject {
+        availableCameras[.front] = deviceManager.cameras.first(where: { $0.cameraFacing == .front })
+        availableCameras[.back] = deviceManager.cameras.first(where: { $0.cameraFacing == .back })
+
+        if let frontCamera = availableCameras[.front] {
```
Do we need to force the localVideoStream to be the front Camera?

cheukchuen replied:
Yes, it was front by default too. Previous implementation `deviceManager.cameras[0]` gets the front camera.

jimchou-dev reviewed (dismissed):
Tested, LGTM

cheukchuen commented on AzureCalling/Controllers/CallViewController.swift:
```diff
@@ -312,6 +311,10 @@ class CallViewController: UIViewController, UICollectionViewDelegate, UICollecti
         infoHeaderView.toggleDisplay()
     }
 
+    @IBAction func switchCamera(_ sender: UIButton) {
```
This is one way I figured to invoke callingContext's method.
We also need to enable this feature in lobbyView as the next work item.

AmyL219 replied:
Let's use closure/delegate to pass the action from participant view rather than direct linking participant view's button to callingView. 
Please correct me if you have any good reference resource with better approach

cheukchuen replied:
Closure is perfectly fine in this scenario.

cheukchuen committed:
Handle error case when the other camera is absent

cheukchuen committed:
Use argument label for updateCameraSwitch

cheukchuen committed:
Use closure to allow camera switch for local participant

AmyL219 commented on AzureCalling/Controllers/CallViewController.swift:
```diff
@@ -64,6 +64,14 @@ class CallViewController: UIViewController, UICollectionViewDelegate, UICollecti
+            }
+
+            self.callingContext.switchCamera { _ in }
+        }
```
How about the localParticipantView in the grid? Do we need to update the switchCamera closure on the onQueueParticipantViewUpdate() function?

cheukchuen replied:
No need. We have only a single instance of localParticipantView, so once we assign a proper switchCamera closure to localParticipantView it'll be functional regardless of the grid configuration (I've tested it).
Remote participants will still have the empty non-functional closure for switchCamera that does nothing.

waynemo commented on AzureCalling/Views/ParticipantView.swift:
```diff
@@ -12,6 +12,9 @@ class ParticipantView: UIView {
 
     // MARK: Properties
 
+    var switchCamera: () -> Void = {}
```
Instead of a property, it might be better to have a method to set the callback:
`setOnSwitchCamera()`

cheukchuen replied:
Meaning we prefer private fields to limit access and have a setter method instead? Got it.

waynemo commented on AzureCalling/Views/ParticipantView.swift:
```diff
@@ -37,6 +42,12 @@ class ParticipantView: UIView {
 
+    // MARK: Actions
+
+    @IBAction func switchCamera(_ sender: UIButton) {
```
Since switch camera functionality is only available for local participants, we should consider a separate class for `LocalParticipantView` that extends `ParticipantView`.

cheukchuen replied:
Good point! That should abstract away some of the logic (including video toggling) pertaining to local participant activities.

cheukchuen committed:
Dedicated class for local participant view

Gemakk reviewed (approved):
A quick triple tap using a real device leave the camera grey.  For remote participants, they are notified that video is on, but nothing appears (Also grey).
![image](https://user-images.githubusercontent.com/6021109/120531089-197a8200-c393-11eb-9109-a1294062c1c7.png)

Gemakk reviewed (changes requested):
See comments above

cheukchuen committed:
Safety feature to disable button while processing camera switch

Gemakk reviewed (dismissed):
Tested

waynemo commented on AzureCalling/Views/ParticipantView.swift:
```diff
@@ -20,9 +20,11 @@ class ParticipantView: UIView {
     // MARK: IBOutlets
 
     @IBOutlet var contentView: UIView!
+    @IBOutlet weak var containerView: UIView!
     @IBOutlet weak var placeholderImage: UIImageView!
     @IBOutlet weak var videoViewContainer: UIView!
     @IBOutlet weak var activeSpeakerView: ActiveSpeakerView!
+    @IBOutlet weak var switchCameraButton: UIRoundedButton!
```
Should these newly added properties only belong to LocalParticipantView?

cheukchuen replied:
They're referencing outlets in the ParticipantView.xib storyboard. Can't seem to connect those properties if we move them to LocalParticipantView, unless we duplicate the storyboard into LocalParticipantView.xib and remove any camera switch elements in ParticipantView.xib.

@AmyL219 you have any idea how to connect the outlets outside of the File Owner's class without duplicating the storyboard? Or that's the only option to have a clear separation of concerns.

cheukchuen replied:
Programmatically configure the button in LocalParticipantView instead of relying on ParticipantView.xib with the latest commit. That should remove any references to camera switch in ParticipantView

cheukchuen committed:
Move any reference of switchCameraButton to LocalParticipantView

- Programmatically configure the button in LocalParticipantView instead of relying on ParticipantView.xib

cheukchuen committed:
Merge branch 'main' into csiow/camera-switching
